File: WSHubsAPI/examples/Clients/_static/WSHubsApi.py
import json
import logging
import threading
from ws4py.client.threadedclient import WebSocketClient
from threading import Timer
from datetime import datetime
log = logging.getLogger(__name__)

class WSSimpleObject(object):
    def __setattr__(self, key, value):
        return super(WSSimpleObject, self).__setattr__(key, value)

class WSReturnObject:
    class WSCallbacks:
        def __init__(self, onSuccess=None, onError=None):
            self.onSuccess = onSuccess
            self.onError = onError

    def done(self, onSuccess, onError=None):
        pass

class GenericServer(object):
    __messageID = 0
    __messageLock = threading.RLock()

    def __init__(self, wsClient, hubName):
        """
        :type wsClient: WSHubsAPIClient
        """
        self.wsClient = wsClient
        self.hubName = hubName

    @classmethod
    def _getNextMessageID(cls):
        with cls.__messageLock:
            cls.__messageID += 1
            return cls.__messageID

    @classmethod
    def _serializeObject(cls, obj2ser):
        obj = obj2ser if not hasattr(obj2ser, "__dict__") else obj2ser.__dict__
        if isinstance(obj,dict):
            sObj = {}
            for key, value in obj.items():
                if isinstance(value,datetime):
                    sObj[key] = value.strftime('%Y/%m/%d %H:%M:%S %f')
                else:
                    try:
                        if not key.startswith("_") and id(value) != id(obj2ser):
                            sValue = cls._serializeObject(value)
                            json.dumps(cls._serializeObject(sValue))
                            sObj[key] = sValue
                    except TypeError:
                        pass
        elif isinstance(obj,(list,tuple,set)):
            sObj = []
            for value in obj:
                try:
                    sValue = cls._serializeObject(value)
                    json.dumps(sValue)
                    sObj.append(sValue)
                except TypeError:
                    pass
        else:
            sObj = obj
        return sObj


class WSHubsAPIClient(WebSocketClient):
    def __init__(self, api, url, serverTimeout):
        super(WSHubsAPIClient, self).__init__(url)
        self.__returnFunctions = dict()
        self.isOpened = False
        self.serverTimeout = serverTimeout
        self.api = api

    def opened(self):
        self.isOpened = True
        log.debug("Connection opened")

    def closed(self, code, reason=None):
        log.debug("Connection closed with code:\n%s\nAnd reason:\n%s" % (code, reason))

    def received_message(self, m):
        try:
            msgObj = json.loads(m.data.decode('utf-8'))
            if "replay" in msgObj:
                f = self.__returnFunctions.get(msgObj["ID"], None)
                if f and msgObj["success"]:
                    f.onSuccess(msgObj["replay"])
                elif f and f.onError:
                    f.onError(msgObj["replay"])
            else:
                self.api.__getattribute__(msgObj["hub"]).client.__dict__[msgObj["function"]](*msgObj["args"])
        except Exception as e:
            self.onError(e)

    def getReturnFunction(self, ID):
        """
        :rtype : WSReturnObject
        """

        def returnFunction(onSuccess, onError=None):
            callBacks = self.__returnFunctions.get(ID, WSReturnObject.WSCallbacks())

            def onSuccessWrapper(*args, **kwargs):
                onSuccess(*args, **kwargs)
                self.__returnFunctions.pop(ID, None)

            callBacks.onSuccess = onSuccessWrapper
            if onError is not None:
                def onErrorWrapper(*args, **kwargs):
                    onError(*args, **kwargs)
                    self.__returnFunctions.pop(ID, None)

                callBacks.onError = onErrorWrapper
            else:
                callBacks.onError = None
            self.__returnFunctions[ID] = callBacks
            r = Timer(self.serverTimeout, self.onTimeOut, (ID,))
            r.start()

        retObject = WSReturnObject()
        retObject.done = returnFunction

        # todo create timeout
        return retObject

    def onError(self, exception):
        log.exception("Error in protocol")

    def onTimeOut(self, messageId):
        f = self.__returnFunctions.pop(messageId, None)
        if f and f.onError:
            f.onError("timeOut Error")

class HubsAPI(object):
    def __init__(self, url, serverTimeout=5.0):
        self.wsClient = WSHubsAPIClient(self, url, serverTimeout)
        self.ChatHub = self.__ChatHub(self.wsClient)

    def connect(self):
        self.wsClient.connect()


    class __ChatHub(object):
        def __init__(self, wsClient):
            hubName = self.__class__.__name__[2:]
            self.server = self.__Server(wsClient, hubName)
            self.client = WSSimpleObject()

        class __Server(GenericServer):
            
            def getNumOfClientsConnected(self, ):
                """
                :rtype : WSReturnObject
                """
                args = list()
                
                id = self._getNextMessageID()
                body = {"hub": self.hubName, "function": "getNumOfClientsConnected", "args": args, "ID": id}
                self.wsClient.send(json.dumps(self._serializeObject(body)))
                return self.wsClient.getReturnFunction(id)
        
            def sendToAll(self, name, message):
                """
                :rtype : WSReturnObject
                """
                args = list()
                args.append(name)
                args.append(message)
                id = self._getNextMessageID()
                body = {"hub": self.hubName, "function": "sendToAll", "args": args, "ID": id}
                self.wsClient.send(json.dumps(self._serializeObject(body)))
                return self.wsClient.getReturnFunction(id)
        

File: wshubsapi/examples/Clients/_static/WSHubsApi.js
/* jshint ignore:start */
/* ignore jslint start */
function HubsAPI(url, serverTimeout) {
    'use strict';

    var messageID = 0,
        returnFunctions = {},
        defaultRespondTimeout = (serverTimeout || 5) * 1000,
        thisApi = this,
        messagesBeforeOpen = [],
        onOpenTriggers = [];
    url = url || '';

    this.clearTriggers = function () {
        messagesBeforeOpen = [];
        onOpenTriggers = [];
    };

    this.connect = function (reconnectTimeout) {
        reconnectTimeout = reconnectTimeout || -1;
        var openPromise = {
            onSuccess : function() {},
            onError : function(error) {}
        };
        function reconnect(error) {
            if (reconnectTimeout !== -1) {
                window.setTimeout(function () {
                    thisApi.connect(reconnectTimeout);
                    thisApi.callbacks.onReconnecting(error);
                }, reconnectTimeout * 1000);
            }
        }

        try {
            this.wsClient = new WebSocket(url);
        } catch (error) {
            reconnect(error);
            return;
        }

        this.wsClient.onopen = function () {
            openPromise.onSuccess();
            openPromise.onError = function () {};
            thisApi.callbacks.onOpen(thisApi);
            onOpenTriggers.forEach(function (trigger) {
                trigger();
            });
            messagesBeforeOpen.forEach(function (message) {
                thisApi.wsClient.send(message);
            });
        };

        this.wsClient.onclose = function (error) {
            openPromise.onError(error);
            thisApi.callbacks.onClose(error);
            reconnect(error);
        };

        this.wsClient.addOnOpenTrigger = function (trigger) {
            if (thisApi.wsClient.readyState === 0) {
                onOpenTriggers.push(trigger);
            } else if (thisApi.wsClient.readyState === 1) {
                trigger();
            } else {
                throw new Error("web socket is closed");
            }
        };

        this.wsClient.onmessage = function (ev) {
            try {
                var f,
                msgObj = JSON.parse(ev.data);
                if (msgObj.hasOwnProperty('replay')) {
                    f = returnFunctions[msgObj.ID];
                    if (msgObj.success && f !== undefined && f.onSuccess !== undefined) {
                        f.onSuccess(msgObj.replay);
                    }
                    if (!msgObj.success) {
                        if (f !== undefined && f.onError !== undefined) {
                            f.onError(msgObj.replay);
                        }
                    }
                } else {
                    f = thisApi[msgObj.hub].client[msgObj.function];
                    if (f!== undefined) {
                        f.apply(f, msgObj.args);
                    } else {
                        this.onClientFunctionNotFound(msgObj.hub, msgObj.function);
                    }
                }
            } catch (err) {
                this.onMessageError(err);
            }
        };

        this.wsClient.onMessageError = function (error) {
            thisApi.callbacks.onMessageError(error);
        };

        return { done: function (onSuccess, onError) {
                openPromise.onSuccess = onSuccess;
                openPromise.onError = onError;
            }
        };
    };

    this.callbacks = {
        onClose: function (error) {},
        onOpen: function () {},
        onReconnecting: function () {},
        onMessageError: function (error){},
        onClientFunctionNotFound: function (hub, func) {}
    };

    this.defaultErrorHandler = null;

    var constructMessage = function (hubName, functionName, args) {
        if(thisApi.wsClient === undefined) {
            throw Error('ws not connected');
        }
        args = Array.prototype.slice.call(args);
        var id = messageID++,
            body = {'hub': hubName, 'function': functionName, 'args': args, 'ID': id};
        if(thisApi.wsClient.readyState === WebSocket.CONNECTING) {
            messagesBeforeOpen.push(JSON.stringify(body));
        } else if (thisApi.wsClient.readyState !== WebSocket.OPEN) {
            window.setTimeout(function () {
                var f = returnFunctions[id];
                if (f !== undefined && f.onError !== undefined) {
                    f.onError('webSocket not connected');
                }
            }, 0);
            return {done: getReturnFunction(id, {hubName: hubName, functionName: functionName, args: args})};
        }
        else {
            thisApi.wsClient.send(JSON.stringify(body));
        }
        return {done: getReturnFunction(id, {hubName: hubName, functionName: functionName, args: args})};
    };
    var getReturnFunction = function (ID, callInfo) {
        return function (onSuccess, onError, respondsTimeout) {
            if (returnFunctions[ID] === undefined) {
                returnFunctions[ID] = {};
            }
            var f = returnFunctions[ID];
            f.onSuccess = function () {
                if(onSuccess !== undefined) {
                    onSuccess.apply(onSuccess, arguments);
                }
                delete returnFunctions[ID];
            };
            f.onError = function () {
                if(onError !== undefined) {
                    onError.apply(onError, arguments);
                } else if (thisApi.defaultErrorHandler !== null){
                    var argumentsArray = [callInfo].concat(arguments);
                    thisApi.defaultErrorHandler.apply(thisApi.defaultErrorHandler.apply, argumentsArray);
                }
                delete returnFunctions[ID];
            };
            //check returnFunctions, memory leak
            respondsTimeout = undefined ? defaultRespondTimeout : respondsTimeout;
            if(respondsTimeout >=0) {
                setTimeout(function () {
                    if (returnFunctions[ID] && returnFunctions[ID].onError) {
                        returnFunctions[ID].onError('timeOut Error');
                    }
                }, defaultRespondTimeout);
            }
        };
    };

    
    this.ChatHub = {};
    this.ChatHub.server = {
        __HUB_NAME : 'ChatHub',
        
        classMethod : function (){
            
            return constructMessage('ChatHub', 'classMethod', arguments);
        },

        staticFunc : function (){
            
            return constructMessage('ChatHub', 'staticFunc', arguments);
        },

        sendToAll : function (name, message){
            arguments[0] = name === undefined ? "hello" : name;
            return constructMessage('ChatHub', 'sendToAll', arguments);
        },

        getSubscribedClientsToHub : function (){
            
            return constructMessage('ChatHub', 'getSubscribedClientsToHub', arguments);
        },

        unsubscribeFromHub : function (){
            
            return constructMessage('ChatHub', 'unsubscribeFromHub', arguments);
        },

        subscribeToHub : function (){
            
            return constructMessage('ChatHub', 'subscribeToHub', arguments);
        }
    };
    this.ChatHub.client = {};
    this.UtilsAPIHub = {};
    this.UtilsAPIHub.server = {
        __HUB_NAME : 'UtilsAPIHub',
        
        getSubscribedClientsToHub : function (){
            
            return constructMessage('UtilsAPIHub', 'getSubscribedClientsToHub', arguments);
        },

        getId : function (){
            
            return constructMessage('UtilsAPIHub', 'getId', arguments);
        },

        isClientConnected : function (clientId){
            
            return constructMessage('UtilsAPIHub', 'isClientConnected', arguments);
        },

        unsubscribeFromHub : function (){
            
            return constructMessage('UtilsAPIHub', 'unsubscribeFromHub', arguments);
        },

        subscribeToHub : function (){
            
            return constructMessage('UtilsAPIHub', 'subscribeToHub', arguments);
        },

        setId : function (clientId){
            
            return constructMessage('UtilsAPIHub', 'setId', arguments);
        },

        getHubsStructure : function (){
            
            return constructMessage('UtilsAPIHub', 'getHubsStructure', arguments);
        }
    };
    this.UtilsAPIHub.client = {};
}
/* jshint ignore:end */
/* ignore jslint end */
    